### PR TITLE
Quick tour

### PR DESCRIPTION
**Pull request by sgugger**
This PR replaces the old quickstart guide by splitting it in two:
- a quick tour that is before the installation page,
- the philosophy page, after the installation.

It also renames Usage to Task summary since it's really what it is, and renames the files usage to task_summary, summary to model_summary.

The quick tour tries to go over all the API introduced in Transformers at a high level, pointing out tutorials (existing or upcoming) for each of them.

### CODE REVIEW

File: docs/source/philosophy.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+- Be as easy and fast to use as possible:
+
+  - We strongly limited the number of user-facing abstractions to learn, in fact, there are almost no abstractions,
+    just three standard classes required to use each model: configuration, models and tokenizer.
+  - All of these classes can be initialized in a simple and unified way from pretrained instances by using a common
+    `from_pretrained()` instantiation method which will take care of downloading (if needed), caching and loading the
+    related class from a pretrained instance supplied in the library or your own saved instance.
+  - On top of those common abstractions for the model, the library provdes two APIs: `pipeline` for quickly using a
+    model on a given task and `Trainer`/`TFTrainer` to quickly train or fine-tune a given model.
```
Comment (thomwolf) [inline docs/source/philosophy.md:19]:
```suggestion
    model (plus its associated tokenizer and configuration) on a given task and `Trainer`/`TFTrainer` to quickly train or fine-tune a given model.
```

File: docs/source/philosophy.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  - As a consequence, this library is NOT a modular toolbox of building blocks for neural nets. If you want to
+    extend/build-upon the library, just use regular Python/PyTorch modules and inherit from the base classes of the
```
Comment (thomwolf) [inline docs/source/philosophy.md:21]:
```suggestion
    extend/build-upon the library, just use regular Python/PyTorch/TensorFlow/Keras modules and inherit from the base classes of the
```

File: docs/source/philosophy.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  - A simple/consistent way to add new tokens to the vocabulary and embeddings for fine-tuning.
+  - Simple ways to mask and prune transformer heads.
+
+- Switch easily between PyTorch and TensorFlow 2.0, allowing training using one framwork and inference using another.
+
+## Main concepts
+
+The library is build around three types of classes for each model:
+
+- **Model classes**  e.g., `BertModel` which are 20+ PyTorch models (`torch.nn.Modules`) that work with the pretrained
```
Comment (thomwolf) [inline docs/source/philosophy.md:49]: link to https://pytorch.org/docs/stable/nn.html#module

File: docs/source/philosophy.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  weights provided in the library. In TensorFlow 2.0, these are `tf.keras.Model`.
```
Comment (thomwolf) [inline docs/source/philosophy.md:50]: link to https://www.tensorflow.org/api_docs/python/tf/keras/Model

File: docs/source/philosophy.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+- `from_pretrained()` let you instantiate a model/configuration/tokenizer from a pretrained version either provided by
+  the library itself (the suported models are provided in the list [here](https://huggingface.co/transformers/pretrained_models.html))
+  or stored locally (or on a server) by the user,
+- `save_pretrained()` let you save a model/configuration/tokenizer locally so that it can be reloaded using
+  `from_pretrained()`.
+
+We'll finish this quickstart tour by going through a few simple quick-start examples to see how we can instantiate and use these classes. The rest of the documentation is organized into two parts:
+
+- the **MAIN CLASSES** section details the common functionalities/method/attributes of the three main type of classes (configuration, model, tokenizer) plus some optimization related classes provided as utilities for training,
+- the **PACKAGE REFERENCE** section details all the variants of each class for each model architectures and, in particular, the input/output that you should expect when calling each of them.
```
Comment (thomwolf) [inline docs/source/philosophy.md:69]: Is it still the case?

**Review by thomwolf — APPROVED** [review]
Very impressive work @sgugger!

File: docs/source/quicktour.rst [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    outputs = model(**batch, labels = torch.tensor([1, 0])
+    ## TENSORFLOW CODE
+    import tensorflow as tf
+    outputs = model(batch, labels = tf.constant([1, 0])
+
+Models are standard `torch.nn.Module <https://pytorch.org/docs/stable/nn.html#torch.nn.Module>`__ or
+`tf.keras.Model <https://www.tensorflow.org/api_docs/python/tf/keras/Model>`__ so you can use them in your usual
+training loop. 🤗 Transformers also provides a :class:`~transformers.Trainer` (or :class:`~transformers.TFTrainer` if
+you are using TensorFlow) class to help with your training (taking care of things such as distributed training, mixed
+precision, etc.). See the training tutorial (coming soon) for more details.
```
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.rst:261]: That's @joeddav tutorial in #5034

File: docs/source/quicktour.rst [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+information about it. It uses the :doc:`DistilBERT architecture </model_doc/distilbert>` and has been fine-tuned on a
+dataset called SST-2 for the sentiment analysis task.
+
+Let's say we want to use another model; for instance, one that has been trained on French data. We can search through
+the `model hub <https://huggingface.co/models>`__ that gathers models pretrained on a lot of data by research labs, but
+also community models (usually fine-tuned versions of those big models on a specific dataset). Applying the tags
+"French" and "text-classification" gives back a suggestion "nlptown/bert-base-multilingual-uncased-sentiment". Let's
+see how we can use it.
+
+We will need two classes for this. The first is :class:`~transformers.AutoTokenizer`, which we will use to download the
```
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.rst:84]: Oh I had missed that. The `AutoXXX` is to transition to the next part, but I can definitely add this example before as another step.

File: docs/source/quicktour.rst [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+.. note::
+
+    All code examples presented in the documentation have a switch on the top left for Pytorch versus TensorFlow. If
+    not, the code is expected to work for both backends without any change needed.
+
+Getting started on a task with a pipeline
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The easiest way to use a pretrained model on a given task is to use :func:`~transformers.pipeline`. 🤗 Transformers
```
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.rst:19]: If you mean the task summary, it's linked just below (this will link to the pipeline doc already).

File: docs/source/quicktour.rst [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+Let's say we want to use another model; for instance, one that has been trained on French data. We can search through
+the `model hub <https://huggingface.co/models>`__ that gathers models pretrained on a lot of data by research labs, but
+also community models (usually fine-tuned versions of those big models on a specific dataset). Applying the tags
+"French" and "text-classification" gives back a suggestion "nlptown/bert-base-multilingual-uncased-sentiment". Let's
+see how we can use it.
+
+We will need two classes for this. The first is :class:`~transformers.AutoTokenizer`, which we will use to download the
+tokenizer associated to the model we picked and instantiate it. The second is
+:class:`~transformers.AutoModelForSequenceClassification` (or
+:class:`~transformers.TFAutoModelForSequenceClassification` if you are using TensorFlow), which we will use to download
```
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.rst:87]: Should I add them to the model_doc/auto file? That would be consistent with the other model pages (I'll do this in a separate PR if yes).

File: docs/source/quicktour.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/quicktour.rst:261]: We should have more easy to use training tutorials, especially for seq2seq models. There are sooo many questions on github asking for T5, Bart and EncoderDecoder training. I added some sample code to the T5 code page, but I like @joeddav approach of having a unified training page where we should at least for `MaskedLM`, `CausalLM` and `Seq2Seq` models how to train in like 10 lines and on top of that how to use the trainer. 

Also at the moment it is not possible to use TFTrainer with `BertForMaskedLM`, `BertForCausalLM` since the forward function does not yet accept a `labels` parameter. I'd be happy to work with @jplu on this in a couple of days :-)

File: docs/source/quicktour.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/quicktour.rst:87]: Agree very much!

**Comment by patrickvonplaten** [comment]
That's awesome, I think this PR is really cool! I wonder whether A short section showing a super simple training example should already be in the quicktour

**Review by patrickvonplaten — APPROVED** [review]

**Comment by sgugger** [comment]
For the quick training example, I have it in mind but for when the integration with nlp is smooth: gathering the data takes a bit too long in code right now, and we have to talk about some obscure function that prepares it, which doesn't really fit the rest.

File: docs/source/quicktour.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (jplu) [reply docs/source/quicktour.rst:261]: I will be happy to work with you on making the `TFTrainer` compliant for language modeling :)

File: docs/source/quicktour.rst [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+To apply these steps on a given text, we can just feed it to our tokenizer:
+
+::
+
+    input = tokenizer("We are very happy to show you the Transformers library.")
+    print(input)
+
+This returns a dictionary string to list of ints. It contains the `ids of the tokens <glossary.html#input-ids>`__,
+as mentioned before, but also additional arguments that will be useful to the model. Here for instance, we also have an
+`attention mask <glossary.html#attention-mask>`__ because the model will expect it:
```
Comment (LysandreJik) [inline docs/source/quicktour.rst:171]: nitpick: The `attention_mask` part here sounds a bit weird to me, as without clicking the link, I don't understand the purpose of that at all (especially since it's the first reference to the attention mask in the docs). I think something along the lines of:
```suggestion
as mentioned before, but also additional arguments that will be useful to the model. Here for instance, we also have an
`attention mask <glossary.html#attention-mask>`__ that the model will use to have a better understanding of the sequence:

```
would be better.

**Review by LysandreJik — APPROVED** [review]
This is really really nice!

File: docs/source/quicktour.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (thomwolf) [reply docs/source/quicktour.rst:19]: I'm not sure it links to the pipeline doc, maybe good to double check

[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.rst:19]: Eh, that's because the pipeline function is not included in the doc... Will fix.

File: docs/source/philosophy.md (hunk 5 above)
Comment (sgugger) [reply docs/source/philosophy.md]: Nope, removed and added a brief summary of the five new sections in the index.

File: docs/source/quicktour.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/quicktour.rst:19]: Fixed now
